packages/core/src/pubsub/task-queue.ts

```diff
@@ -0,0 +1,43 @@
+import PQueue from 'p-queue';
+
+const noop = () => {
+  // Do Nothing
+};
+
+type Task<TaskResultType> = (() => PromiseLike<TaskResultType>) | (() => TaskResultType);
+
+/**
+ * PQueue with synchronous `add` and a common error-handler.
+ */
+export class TaskQueue {
+  #pq = new PQueue({ concurrency: 1 });
+
+  /**
+   * @param onError - Common error handler.
+   */
+  constructor(private readonly onError: (error: Error, retry?: () => void) => void = noop) {}
+
+  /**
+   * Add task to queue.
+   */
+  add(f: Task<void>): void {
+    this.#pq.add(f).catch((error) => {
+      const retry = () => this.add(f);
+      this.onError(error, retry);
```

Comment:
does this mean that all tasks in the queue that error will be retried indefinitely?  Should we have a cap of how many retry attempts it gets so we don't get stuck in an infinitely failing loop?

EDIT: Nevermind, I see now it's the `onError` callback that decides if/when to retry.  I think we need some comments to make that clearer.  Maybe give the onError function signature a type and have a comment describing how it's used?

Reply:
I'd still love some more comments or something here as I think the relationship between the `onError` callback given to the constructor and the `retry` argument it receives isn't super obvious at first glance

Reply:
Ack. Hope now it is more clear.